### ISSUE TITLE
Persistence cannot remove files

### ISSUE THREAD

**Issue by dismirlian** [eclipse-paho/paho.mqtt.c#1035]
**Describe the bug**
Persistence implementation has inconsitencies.

MQTTPersistence.h has the following lines:
```C

/** Stem of the key for a sent PUBLISH QoS1 or QoS2 */
#define PERSISTENCE_PUBLISH_SENT "s-"
/** Stem of the key for a sent PUBREL */
#define PERSISTENCE_PUBREL "sc-"
/** Stem of the key for a received PUBLISH QoS2 */
#define PERSISTENCE_PUBLISH_RECEIVED "r-"

/** Stem of the key for a sent MQTT V5 PUBLISH QoS1 or QoS2 */
#define PERSISTENCE_V5_PUBLISH_SENT "s5-"
/** Stem of the key for a sent MQTT V5 PUBREL */
#define PERSISTENCE_V5_PUBREL "sc5-"
/** Stem of the key for a received MQTT V5 PUBLISH QoS2 */
#define PERSISTENCE_V5_PUBLISH_RECEIVED "r5-"

/** Stem of the key for an async client command */
#define PERSISTENCE_COMMAND_KEY "c-"
/** Stem of the key for an MQTT V5 async client command */
#define PERSISTENCE_V5_COMMAND_KEY "c5-"
/** Stem of the key for an client incoming message queue */
#define PERSISTENCE_QUEUE_KEY "q-"
/** Stem of the key for an MQTT V5 incoming message queue */
#define PERSISTENCE_V5_QUEUE_KEY "q5-"
/** Maximum length of a stem for a persistence key */
#define PERSISTENCE_MAX_STEM_LENGTH 3
/** Maximum allowed length of a persistence key */
#define PERSISTENCE_MAX_KEY_LENGTH 9
/** Maximum size of an integer sequence number appended to a persistence key */
#define PERSISTENCE_SEQNO_LIMIT 1000000 /*10^(PERSISTENCE_MAX_KEY_LENGTH - PERSISTENCE_MAX_STEM_LENGTH)*/
```

`PERSISTENCE_V5_PUBREL` is 4 chars long. IMHO `PERSISTENCE_MAX_STEM_LENGTH` should be 4 and `PERSISTENCE_SEQNO_LIMIT` should be 100k instead of 1M.

Also, and more importantly, the library cannot remove s(c)(5)-10000.msg onwards, because MQTTPersistanceDefault.h defines `MESSAGE_FILENAME_LENGTH` as 8; `MQTTPersistence_remove` cannot print "sc5-10000" in 8 characters, and fails with "Error writing 9 chars with snprintf".

https://github.com/eclipse/paho.mqtt.c/blob/317fb008e1541838d1c29076d2bc5c3e4b6c4f53/src/MQTTPersistence.c#L533-L560

**Comment by fpagliughi** [comment]
Oh, boy, this sounds like the bug I reported in #977! 
(Buffer overflow crash when we reach 10,000 messages with persistence turned on.)
Well at least by converting to `snprintf()` we see the error reported instead of a hard segfault.

**Comment by dismirlian** [comment]
Hi @fpagliughi yes, it definitely seems related. I patched the library locally and was able to work around the problem; however I don't know if it's the best approach. Basically I set:

PERSISTENCE_MAX_STEM_LENGTH=4
PERSISTENCE_SEQNO_LIMIT=100000
MESSAGE_FILENAME_LENGTH=9

**Comment by dismirlian** [comment]
It seems to me that `MESSAGE_FILENAME_LENGTH` and `PERSISTENCE_MAX_KEY_LENGTH` should be the same `#define`.

**Comment by fpagliughi** [comment]
@dismirlian, after looking a little more closely at this, I think you're completely right. 

We see this in Persistence.c a few times:
```
const size_t keysize = MESSAGE_FILENAME_LENGTH + 1;
```
Why is the base persistence code using a #define from the file/default persistence when it has its own #define for this buffer size as `PERSISTENCE_MAX_KEY_LENGTH` ? That isn't logical.

And then why does it define the filename length without basing it on the size of the key?

As for the `seqno`, I'm not sure if it's supposed to be just big enough for the max MQTT message ID (64k = 5 digits) or if they wanted an extra digit to give plenty of extra possible sequence numbers. I wouldn't want to remove the extra digit without knowing the answer to that. 

So I would probably:
- Keep the `PERSISTENCE_SEQNO_LIMIT` at 1M 
- Bump the key length,  `PERSISTENCE_MAX_KEY_LENGTH`, to 10
- Get rid of the `MESSAGE_FILENAME_LENGTH` completely and use the key length

**Comment by fpagliughi** [comment]
@icraggs I can probably do a PR over the weekend for this.

**Comment by icraggs** [comment]
The fix for #977 was applied in the fix for that issue.  I reproduced it and fixed it.

PERSISTENCE_SEQNO_LIMIT applies to the sequence numbers used in the persistence queue.  It's not the MQTT message id. Even though I advise against storing large numbers of messages in such a simple implementation (it's not a database!) some people still seem intent on doing so. 

I assume that this is by inspection rather than testing?  Or do you have a test? I think this needs MQTT V5 QoS 2 to be used beyond the 10000 sequence number.  I'd like to have a test to prove a fix works.

**Comment by fpagliughi** [comment]
It seems that #977 had two underlying problems. You *definitely* found the first one. But this is a problem, too.

Several things seem _clearly_ wrong on inspection of the code as @dismirlian pointed out:

The const `PERSISTENCE_MAX_STEM_LENGTH` is 3, but a new, longer, stem was added for MQTT v5:
```
/** Stem of the key for a sent MQTT V5 PUBREL */
#define PERSISTENCE_V5_PUBREL "sc5-"
```
This is 4 chars long, so the max stem length constant should have been bumped up to 4 when this was added.

And so, to keep a 6-digit seqno (which seems fine), the max key length must be bumped up by one (from 9 to 10 chars):
```
/** Maximum allowed length of a persistence key */
#define PERSISTENCE_MAX_KEY_LENGTH 10
```

But then, _none_ of the code seemed to be using this #define as the size when allocating a buffer to hold a key!
Instead it was using a #define in _MQTTPersistenceDefault.h:_
```
/** 8.3 filesystem */
#define MESSAGE_FILENAME_LENGTH 8    
```
as in:
```
const size_t keysize = MESSAGE_FILENAME_LENGTH + 1;
char *key = malloc(keysize);
...
publish->MQTTVersion = c->MQTTVersion;
if (publish->MQTTVersion >= MQTTVERSION_5)
    chars = snprintf(key, keysize, "%s%d", PERSISTENCE_V5_PUBREL, publish->msgId);
```
This is _clearly_ two bytes too short to hold the maximum key value!

It should be:
```
const size_t keysize = PERSISTENCE_MAX_KEY_LENGTH + 1;
```

**Comment by fpagliughi** [comment]
OK, technically, the new, longer, stem is being used with a `msgId`, not a `seqno`. So I guess practically speaking the longest  one of these would be 4(stem) + 5(digits) = 9byte key. But I think using that as the max key size would be confusing and dangerous. What one extra byte among friends?

But clearly, the moral of the story here is helpful it was that you switched the `sprintf()` calls to `snprintf()`. Although a bug was still in the code, the `snprintf()` kept the app from crashing, and pointed to the exact problem (a key can be 9 bytes) and solution (increase the buffer size).

**Comment by dismirlian** [comment]
Hi @icraggs, this was caught by testing. I left my application running in a place with erratic cellular coverage and after one month or so I had the persistence directory full of messages with these filenames:
- s-1xxxx.msg
- sc-1xxxx.msg

Paho kept printing `Error writing 9 chars with snprintf`.

I grabbed the contents of the directory and reproduced the problem locally, in my development PC with a solid network connection. Paho would never empty the persistence directory, so I digged deeper and found the issue at hand. @fpagliughi explained the problem better than I did; as a proof-of-concept fix I did:
```C
PERSISTENCE_MAX_STEM_LENGTH=4
PERSISTENCE_SEQNO_LIMIT=100000
MESSAGE_FILENAME_LENGTH=9
```
and Paho was able to submit pending messages and empty the persistence directory. Also,
I'm **not** using MQTTv5, though these messages are either QoS 1 or 2 (I don't know which; I could attach the directory contents).

IMHO the fix proposed by @fpagliughi makes sense to me.

**Comment by icraggs** [comment]
@dismirlian did you say which version you were using?  The fix for #977 was in 1.3.7

**Comment by icraggs** [comment]
> But clearly, the moral of the story here is helpful it was that you switched the `sprintf()` calls to `snprintf()`. Although a bug was still in the code, the `snprintf()` kept the app from crashing, and pointed to the exact problem (a key can be 9 bytes) and solution (increase the buffer size).

Conversely, unless the CI test monitoring can check this outcome, it can mask a CI test failure.  A crash will always be caught.

**Comment by fpagliughi** [comment]
Shouldn't a returned error also be caught?

**Comment by icraggs** [comment]
Yes - but that depends on the error being propagated and there are other ways errors could manifest themselves.  Anyway that's predicated on there also being a CI test for this, so I plan to write one.

**Comment by dismirlian** [comment]
> @dismirlian did you say which version you were using? The fix for #977 was in 1.3.7

@icraggs sorry for the delay. I was using tag v1.3.7 (e047e25d34d53b4b265649144a3cac3b01eee76c).
